### PR TITLE
Llama 2 xla optimized

### PR DESCRIPTION
**Pull request by Liyang90**
This is optimized PT/XLA integration for LLaMA 2 inference.

It contains:
* Updating LLaMA 2 text completion and chat completion example scripts
* Changing device to XLA device or CUDA device depending on the `USE_CUDA` flag
* Changing Tensor Parallelism to use PT/XLA APIs
* Generation optimization
* Model optimization
* Long prompt optimization
* Dynamo
* A new resharding script to shard LLaMA 2 checkpoints to more pieces for different sizes of XLA device clusters.

[TODO] Update docs

### CODE REVIEW

File: llama/generation.py [hunk 1]
```diff
@@ -94,124 +108,175 @@ def build(
...
-        if temperature > 0:
-            probs = torch.softmax(logits[:, -1] / temperature, dim=-1)
-            next_token = sample_top_p(probs, top_p)
+            full_logits = self.model(input_tokens, input_pos_tensor, None)
+            logits = full_logits.index_select(1, output_pos_tensor - input_pos_tensor[0]).squeeze(dim=1)
         else:
-            next_token = torch.argmax(logits[:, -1], dim=-1)
+            logits = self.model(input_tokens, input_pos_tensor, output_pos_tensor)
+        if with_temp:
+            probs = torch.softmax(logits / temperature_tensor, dim=-1)
```
Comment (miladm) [inline llama/generation.py:150]: The source is picks up the last column `logits[:,-1]`. Here we pass the whole `logits` tensor to `softmax`. What am I missing?

File: llama/generation.py [hunk 2]
```diff
@@ -94,124 +108,175 @@ def build(
...
+            next_token = sample_top_p(probs, top_p_tensor)
+        else:
+            next_token = torch.argmax(logits, dim=-1)
```
Comment (miladm) [inline llama/generation.py:153]: ditto

File: llama/generation.py (hunk 1 above)
Comment (Liyang90) [reply llama/generation.py:150]: > The source is picks up the last column `logits[:,-1]`. Here we pass the whole `logits` tensor to `softmax`. What am I missing?

I've optimized the `model.py` code here: https://github.com/pytorch-tpu/llama/blob/42fb3340b52fcbdfde0a36823d2649b567817356/llama/model.py#L410-L412

File: llama/generation.py (hunk 2 above)
Comment (Liyang90) [reply llama/generation.py:153]: I've optimized the `model.py` code here: https://github.com/pytorch-tpu/llama/blob/42fb3340b52fcbdfde0a36823d2649b567817356/llama/model.py#L410-L412

**Review by alanwaketan — APPROVED** [review]
LGTM.

**Review by lsy323 — APPROVED** [review]
LGTM
